proposals/SIMD-03

```diff
@@ -0,0 +1,79 @@
+---
+simd: 0287
+title: Message Headers with Compute Budget Metadata
+authors:
+  - Cavey Cool
+category: Standard
+type: Interface
+status: Idea/Review
+created: 2025-05-20
+feature: N/A
+---
+
+## Summary
+
+We introduce three new (candidate) versions for transaction format aiming at
+reducing the transaction footprint for compute budget instructions.
+
+## Motivation
+
+The use of compute budget instructions has become ubiquitous. These instructions
+currently have a nontrivial serialized footprint and are presently wasteful in
+their implementation. The information in these instructions can be compacted
+significantly, enabling users to fit a bit more data in their transaction payload.
+
+## Detailed Design
+
+### v1: Fixed Fields for Compute Unit Limit & Price
+
+- **Change**: `MessageHeader` is extended to include two new fields:
+  - `compute_unit_price` (u64)
+  - `compute_unit_limit` (u32)
+- **Serialization**: These fields are serialized immediately before the existing
+  three `u8` signature counters.
+
+### v2: Fixed Fields for Compute Unit Limit & Price, Loaded Data & Heap Requests
+
+- **Change**: `MessageHeader` is extended to include four new fields:
+  - `compute_unit_price` (u64)
+  - `compute_unit_limit` (u32)
+  - `loaded_accounts_data_limit` (u32)
+  - `requested_heap_bytes` (u32)
+- **Serialization**: These fields are serialized immediately before the existing
+  three `u8` signature counters.
+
+### v3: Dynamic Header
+
+- **Change**: Introduce a new `ComputeBudgetHeader` struct at the front of the
+    message, containing:
+  - `flags: u8` bitmask indicating which compute budget fields are present.
```

Comment:
what're thoughts on the below? for future proof, is there a world where >8 compute-related operations take place?
```
#[repr(u8)]
enum ComputeBudgetOp {
    ComputeUnitPrice(u64),
    ComputeUnitLimit(u32),
    ...
}

ComputeBudgetHeader {
    // perhaps Vec encoded with u8 length discriminator instead of u64
    ops: Vec<ComputeBudgetOp>
}
```

Reply:
in all seriousness, i think if we're worried about future proofing we can use a 16-bit flag

Reply:
how do you envision combining Option<u32> or Option<u64>? will it be one or the other? 

Options comes with [1 byte discriminator](https://github.com/bincode-org/bincode/blob/trunk/docs/spec.md#options) as well. Are you suggesting dropping that with the "serialize without tags" step 2?

8 + (1 * 8) * num_options
4 /* vec len */ + (1 * 8) * num_options

Reply:
maybe I’m confused on options, that’s for building the message but not encoded in which case what I said above is wrong